module-4/task-3/refactor/refactor.go

```diff
@@ -1,40 +1,58 @@
 package main
 
 import (
+	"bytes"
 	"fmt"
 	"io/ioutil"
 	"os"
 )
 
-func readFile(path string) []byte {
+type OpenFileError struct {
+	err  error
+	file string
+}
+
+type CantReadFileError struct {
+	err  error
+	file string
+}
+
+func (e OpenFileError) Error() string {
+	if e.err != nil {
+		return fmt.Sprintf("Error during file %q opening: %v", e.file, e.err.Error())
+	}
+	return fmt.Sprintf("Unknown error during file %q opening", e.file)
+}
+
+func (e CantReadFileError) Error() string {
+	if e.err != nil {
+		return fmt.Sprintf("Error during file %q reading: %v", e.file, e.err.Error())
+	}
+	return fmt.Sprintf("Unknown error during file %q reading", e.file)
+}
+
+func readFile(path string) ([]byte, error) {
 	fd, err := os.OpenFile(path, os.O_RDONLY, os.ModePerm)
 	if err != nil {
-		panic(err)
```

Comment:
Approach with Error types is also possible, but I think that this is the best solution in this code. Typically`return nil, fmt.Errorf("failed to open file: %w")` enough

Reply:
I have just use better practic according to Alieksei lecture

Reply:
This solution also possible, and it is prefered when you rely on the type or have extra data on the error. But usually, it is overhead to write an extra type for every single error.